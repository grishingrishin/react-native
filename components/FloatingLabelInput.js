import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Animated } from 'react-native';
import styled from 'styled-components/native';

const Container = styled.View`
  position: relative;
`;

const Placeholder = styled.Text`
  font-family: Roboto-Light;
  color: #fff;
`;

const TextInput = styled.TextInput`
  padding: 18px 14px 18px 14px;
  font-size: 16px;
  font-family: Roboto-Light;
  color: #fff;
  border-bottom-width: 1px;
  border-bottom-color: rgba(211, 200, 230, 0.5);
`;

class FloatingLabelInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      animatedIsFocused: new Animated.Value(this.hasInputValue()),
    };
  }

  handleFocus = () => {
    return Animated.timing(this.state.animatedIsFocused, {
      toValue: 1,
      duration: 200
    }).start();
  }

  handleBlur = () => {
    return Animated.timing(this.state.animatedIsFocused, {
      toValue: this.hasInputValue(),
      duration: 200
    }).start();
  }

  hasInputValue = () => this.props.value.length !== 0 ? 1 : 0;

  render() {
    return (
      <Container>
        <Animated.Text
          style={{
            position: 'absolute',
            left: 14,
            top: this.state.animatedIsFocused.interpolate({
              inputRange: [0, 1],
              outputRange: [18, 0],
            }),
            fontSize: this.state.animatedIsFocused.interpolate({
              inputRange: [0, 1],
              outputRange: [22, 12],
            }),
            opacity: this.state.animatedIsFocused.interpolate({
              inputRange: [0, 1],
              outputRange: [1, 0.5],
            }),
          }}
        >
          <Placeholder>{this.props.placeholder}</Placeholder>
        </Animated.Text>
        <TextInput
          value={this.props.value}
          secureTextEntry={this.props.secureTextEntry}
          onChangeText={text => this.props.changeHandle(text)}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
      </Container>
    )
  }
}

FloatingLabelInput.propTypes = {
  value: PropTypes.string.isRequired,
  secureTextEntry: PropTypes.bool.isRequired,
  changeHandle: PropTypes.func.isRequired,
}

FloatingLabelInput.defaultProps = {
  value: '',
  secureTextEntry: false,
}

export default FloatingLabelInput;