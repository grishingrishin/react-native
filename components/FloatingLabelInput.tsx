import React, {Component} from 'react';
import {Animated} from 'react-native';
import styled from 'styled-components/native';

const Container = styled.View`
  position: relative;
`;

const Placeholder = styled.Text`
  font-family: Roboto-Light;
  color: #fff;
`;

const TextInput = styled.TextInput`
  padding: 18px 14px 18px 14px;
  font-size: 16px;
  font-family: Roboto-Light;
  color: #fff;
  border-bottom-width: 1px;
  border-bottom-color: rgba(211, 200, 230, 0.5);
`;

interface FLInputProps {
  value: string;
  placeholder: string;
  secureTextEntry: boolean;
  changeHandle(value: string): void;
}

interface FLInputState {
  animatedIsFocused: any;
}

class FloatingLabelInput extends Component<FLInputProps, FLInputState> {
  static defaultProps = {
    value: '',
    secureTextEntry: false,
  };

  constructor(props: FLInputProps) {
    super(props);

    this.state = {
      animatedIsFocused: new Animated.Value(this.hasInputValue()),
    };
  }

  private handleChange = (value: string): void => {
    this.props.changeHandle(value);
  };

  private handleFocus = (): void => {
    Animated.timing(this.state.animatedIsFocused, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }).start();
  };

  private handleBlur = (): void => {
    Animated.timing(this.state.animatedIsFocused, {
      toValue: this.hasInputValue(),
      duration: 200,
      useNativeDriver: false,
    }).start();
  };

  private hasInputValue = (): number => (this.props.value.length !== 0 ? 1 : 0);

  public render() {
    return (
      <Container>
        <Animated.Text
          style={{
            position: 'absolute',
            left: 14,
            top: this.state.animatedIsFocused.interpolate({
              inputRange: [0, 1],
              outputRange: [18, 0],
            }),
            fontSize: this.state.animatedIsFocused.interpolate({
              inputRange: [0, 1],
              outputRange: [22, 12],
            }),
            opacity: this.state.animatedIsFocused.interpolate({
              inputRange: [0, 1],
              outputRange: [1, 0.5],
            }),
          }}>
          <Placeholder>{this.props.placeholder}</Placeholder>
        </Animated.Text>
        <TextInput
          value={this.props.value}
          secureTextEntry={this.props.secureTextEntry}
          onChangeText={this.handleChange}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
      </Container>
    );
  }
}

export default FloatingLabelInput;
