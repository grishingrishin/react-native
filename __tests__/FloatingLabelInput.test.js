import React from 'react';
import { create } from 'react-test-renderer';
import FloatingLabelInput from '../components/FloatingLabelInput';

describe('<FloatingLabelInput />', () => {
  describe('Jest tests', () => {
    it('renders correctly', () => {
      const tree = create(
        <FloatingLabelInput
          value='Some test text value'
          changeHandle={jest.fn()}
        />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
