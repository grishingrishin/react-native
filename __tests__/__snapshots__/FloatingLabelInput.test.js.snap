// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FloatingLabelInput /> Jest tests renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "position": "relative",
      },
    ]
  }
>
  <Text
    style={
      Object {
        "fontSize": 12,
        "left": 14,
        "opacity": 0.5,
        "position": "absolute",
        "top": 0,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontFamily": "Roboto-Light",
          },
        ]
      }
    />
  </Text>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    secureTextEntry={false}
    style={
      Array [
        Object {
          "borderBottomColor": "rgba(211,200,230,0.5)",
          "borderBottomWidth": 1,
          "color": "#fff",
          "fontFamily": "Roboto-Light",
          "fontSize": 16,
          "paddingBottom": 18,
          "paddingLeft": 14,
          "paddingRight": 14,
          "paddingTop": 18,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value="Some test text value"
  />
</View>
`;
